src/ixdat/calculators/packs.py

```diff
@@ -0,0 +1,559 @@
+        """
+        if self._items is not None:
+            return self._items
+
+        items: List[JsonDict] = []
+        for cal in self._calculators:
+            cls = cal.__class__
+            class_path = f"{cls.__module__}.{cls.__name__}"
+
+            payload = self._serialize_calculator(cal)
+
+            item: JsonDict = {
+                "class_path": class_path,
+                "payload": payload,
+                "hash": _sha256_of_jsonable(payload),
+            }
+
+            # Optional metadata
+            cal_ver = getattr(cal, "version", None)
+            if cal_ver:
+                item["calculator_version"] = str(cal_ver)
+
+            items.append(item)
+
+        self._items = items
+        return items
+
+    def to_json(self) -> JsonDict:
+        """
+        Serialize the pack (including calculators) to a JSON-serializable dict.
+
+        Returns:
+            dict: The full, self-describing pack.
+        """
+        items = self._ensure_items()
+        data: JsonDict = {
+            "type": "ixdat.CalculatorPack",
+            "name": self.name,
+            "schema_version": self.schema_version,
+            "ixdat_version": self.ixdat_version,
+            "created_utc": self.created_utc,
+            "metadata": self._metadata,
+            "calculators": items,
+        }
+        # Also keep a string form for Saveable db column
+        self.json = json.dumps(data, indent=2)
+        return data
+
+    @classmethod
+    def from_json(cls, data: JsonDict) -> "CalculatorPack":
+        """
+        Construct a pack from a JSON dictionary (inverse of `to_json`).
+        """
+        if data.get("type") != "ixdat.CalculatorPack":
+            raise ValueError(f"Unexpected type: {data.get('type')}")
+
+        pack = cls(name=data.get("name"))
+        pack.schema_version = str(data.get("schema_version") or cls.SCHEMA_VERSION)
+        pack.ixdat_version = data.get("ixdat_version")
+        pack.created_utc = data.get("created_utc") or _now_utc_iso()
+        pack._metadata = data.get("metadata", {})
+        pack._items = list(data.get("calculators") or [])
+        # Keep canonical db json string if caller wants to save to DB
+        pack.json = json.dumps(data, indent=2)
+        return pack
+
+    def to_file(self, path: str) -> None:
+        """
+        Write the pack to a JSON file.
+
+        Parameters:
+            path: Output file path (e.g., 'my_setup.ixpack.json').
+        """
+        data = self.to_json()
+        with open(path, "w", encoding="utf-8") as f:
+            json.dump(data, f, indent=2)
+
+    @classmethod
+    def read(
+        cls,
+        path: str,
+    ) -> "CalculatorPack":
+        """
+        Read a pack from a file (`path`)
+
+        Returns:
+            CalculatorPack
+        """
+        with open(path, "r", encoding="utf-8") as f:
+            data = json.load(f)
+        return cls.from_json(data)
+
+    def calculators(self) -> List[Any]:
```

Comment:
It can be a bit confusing that this is a method whereas `Measurement.calculators` is a dictionary. Not sure which is best practice. Would you suggest changing `Measurement.calculators` to a method to match `CalculatorPack.calculators`?